From 3e3c733eaa6fb5b408dabb513fc632878aed2db1 Mon Sep 17 00:00:00 2001
From: Alessandro Astone <alessandro.astone@canonical.com>
Date: Tue, 7 Jul 2026 12:38:49 +0200
Subject: [PATCH] glycin: Partially disable sandbox for snap environment

We cannot run bwrap within snap confinement, but we can still set a seccomp
filter at least.
---
 .../glycin-core/src/api/common.rs             |  1 +
 .../glycin-core/src/sandbox.rs                | 59 ++++++++++++++++++-
 .../missing-sources/glycin-core/src/util.rs   | 28 ++++++++-
 3 files changed, 84 insertions(+), 4 deletions(-)

--- a/glycin-core/src/api/common.rs
+++ b/glycin-core/src/api/common.rs
@@ -29,6 +29,7 @@ pub enum SandboxMechanism {
 impl SandboxMechanism {
     pub async fn detect() -> Self {
         match RunEnvironment::cached().await {
+            RunEnvironment::Snap => Self::NotSandboxed,
             RunEnvironment::SandboxForceDisabled => Self::NotSandboxed,
             RunEnvironment::FlatpakDevel => Self::NotSandboxed,
             RunEnvironment::Flatpak => Self::FlatpakSpawn,
--- a/glycin-core/src/sandbox.rs
+++ b/glycin-core/src/sandbox.rs
@@ -21,13 +21,22 @@ use nix::unistd;
 
 use crate::config::{ConfigEntry, LoaderConfig, Processor};
 use crate::error::ErrorKind;
-use crate::util::{self, AsyncMutex, new_async_mutex, spawn_blocking};
+use crate::util::{self, AsyncMutex, RunEnvironment, new_async_mutex, spawn_blocking};
 use crate::{Error, SandboxMechanism};
 
 type SystemSetupStore = Arc<Result<SystemSetup, Arc<io::Error>>>;
 
 static SYSTEM_SETUP: AsyncMutex<Option<SystemSetupStore>> = new_async_mutex(None);
 
+struct SendableFilterContext(ScmpFilterContext);
+unsafe impl Send for SendableFilterContext {}
+unsafe impl Sync for SendableFilterContext {}
+impl SendableFilterContext {
+    fn load(&self) -> Result<(), libseccomp::error::SeccompError> {
+        self.0.load()
+    }
+}
+
 /**** BEGIN NOTE ON CODE SHARING
  *
  * This code is copied from Flatpak:
@@ -142,6 +151,7 @@ const BLOCKED_SYSCALLS: &[(&str, ScmpAct
 ];
 
 const INHERITED_ENVIRONMENT_VARIABLES: &[&str] = &["RUST_BACKTRACE", "RUST_LOG", "XDG_RUNTIME_DIR"];
+const INHERITED_ENVIRONMENT_VARIABLES_SNAP: &[&str] = &["LD_LIBRARY_PATH"];
 
 pub struct Sandbox {
     sandbox_mechanism: SandboxMechanism,
@@ -208,7 +218,10 @@ impl Sandbox {
                 (command, None)
             }
             SandboxMechanism::NotSandboxed => {
-                let command = self.no_sandbox_command();
+                let command = match RunEnvironment::cached().await {
+                    RunEnvironment::Snap => self.snap_command()?,
+                    _ => self.no_sandbox_command(),
+                };
 
                 (command, None)
             }
@@ -540,6 +553,39 @@ impl Sandbox {
         command
     }
 
+    fn snap_command(&self) -> Result<Command, Error> {
+        let mut command = Command::new(self.exec());
+
+        command.env_clear();
+        for key in INHERITED_ENVIRONMENT_VARIABLES {
+            if let Some(val) = std::env::var_os(key) {
+                command.env(key, val);
+            }
+        }
+        for key in INHERITED_ENVIRONMENT_VARIABLES_SNAP {
+            if let Some(val) = std::env::var_os(key) {
+                command.env(key, val);
+            }
+        }
+
+        let filter = self.seccomp_filter()?;
+        let proc_filter = SendableFilterContext(filter);
+
+        unsafe {
+            command.pre_exec(move || {
+                nix::sys::prctl::set_pdeathsig(nix::sys::signal::SIGKILL).map_err(std::io::Error::from)?;
+                nix::sys::prctl::set_no_new_privs().map_err(std::io::Error::from)?;
+
+                proc_filter.load().or_else(|_| {
+                    libc_eprint("glycin sandbox: Failed to load seccomp filter, continuing without\n");
+                    Ok(())
+                })
+            });
+        }
+
+        Ok(command)
+    }
+
     /// Memory limit in bytes that should be applied to sandboxes
     fn memory_limit() -> resource::rlim_t {
         // Lookup free memory
--- a/glycin-core/src/util.rs
+++ b/glycin-core/src/util.rs
@@ -131,6 +131,8 @@ pub enum RunEnvironment {
     Host,
     /// Not inside Flatpak but bwrap doesn't work
     HostBwrapSyscallsBlocked,
+    /// Inside Snap confinement
+    Snap,
     /// Inside Flatpak
     Flatpak,
     /// Inside Flatpak and development environment
@@ -154,6 +156,11 @@ impl RunEnvironment {
                     "WARNING: Glycin running without sandbox. Force disabled via environment variable."
                 );
                 Self::SandboxForceDisabled
+            } else if snap_confined() {
+                eprintln!(
+                    "WARNING: Glycin running without additional sandbox inside snap confinement."
+                );
+                Self::Snap
             } else if let Some(devel) = flatpak_devel().await {
                 if devel {
                     eprintln!(
@@ -183,6 +190,11 @@ impl RunEnvironment {
     }
 }
 
+#[cfg(feature = "external")]
+fn snap_confined() -> bool {
+    std::env::var_os("SNAP").is_some_and(|val| !val.is_empty())
+}
+
 /// Returns None if not in Flatpak environment, otherwise true if development
 #[cfg(feature = "external")]
 async fn flatpak_devel() -> Option<bool> {
